### ISSUE TITLE
Buildpack does not bind Kestrel to the required URL

### ISSUE THREAD

**Issue by nathan-alden** [cloudfoundry/dotnet-core-buildpack#137]
>What version of Cloud Foundry and CF CLI are you using? (i.e. What is the output of running `cf curl /v2/info && cf version`?
```
{
   "name": "",
   "build": "",
   "support": "https://support.pivotal.io",
   "version": 0,
   "description": "",
   "authorization_endpoint": "<redacted>",
   "token_endpoint": "<redacted>",
   "min_cli_version": "6.7.0",
   "min_recommended_cli_version": "6.11.2",
   "api_version": "2.65.0",
   "app_ssh_endpoint": "<redacted>",
   "app_ssh_host_key_fingerprint": "89:04:08:96:ba:26:ed:ed:5e:0f:b5:e5:6b:5a:73:86",
   "app_ssh_oauth_client": "ssh-proxy",
   "routing_endpoint": "<redacted>",
   "logging_endpoint": "<redacted>",
   "doppler_logging_endpoint": "<redacted>"
}
```
```
cf.exe version 6.25.0+787326d.2017-02-28
```

>What version of the buildpack you are using?

1.0.13

>If you were attempting to accomplish a task, what was it you were attempting to do?

I'm trying to get an ASP.NET Core application running.

>What did you expect to happen?

Kestrel would be bound to the correct URL, which should be `http://0.0.0.0:${PORT}`.

>What was the actual behavior?

Kestrel was bound to the default URL of `http://localhost:5000`.

>Please confirm where necessary:
* [ ] I have included a log output
* [ ] My log includes an error message
* [X] I have included steps for reproduction

Steps to reproduce:
1. Deploy any .NET Core 1.1.1 app, then look at the logs. You will notice Kestrel is bound to `http://localhost:5000` instead of `http://0.0.0.0:${PORT}`.

Analysis:

This is happening because the buildpack runs the app using the no-longer-working `server.urls` command line parameter of `dotnet`. Kestrel now requires that the `ASPNETCORE_URLS` environment variable be used to supply the URL that Kestrel will bind to.

**Comment by degrim** [comment]
@nathan-alden Can you show your `Main` method and how you are configuring Kestrel there?  The `--server.urls` parameter should be working if you're configuring kestrel to use command-line configuration.

**Comment by degrim** [comment]
@sesmith177 @sclevine We should probably auto-set the `ASPNETCORE_URLS` parameter as well as passing the `--server.urls` parameter too.

**Comment by nathan-alden** [comment]
Our bootstrap code does not explicitly set the Kestrel URLs, since we rely on the environment variable:
```
using (IWebHost host = new WebHostBuilder()
                .UseKestrel()
                .UseContentRoot(Directory.GetCurrentDirectory())
                .UseStartup<Startup>()
                .Build())
            {
                host.Run();
            }
```

**Comment by nathan-alden** [comment]
I was told that adding ConfigurationBuilder back into the bootstrap code and reading the command line with it would allow injection of `server.urls`. I'll do this temporarily.

**Comment by degrim** [comment]
@nathan-alden That is correct, you would need to use the `ConfigurationBuilder` as described in the [buildpack documentation](http://docs.cloudfoundry.org/buildpacks/dotnet-core/index.html#port) for now.

@sesmith177 I'll look at replacing the deprecated `--server.urls` parameter with the `ASPNETCORE_URLS` environment variable and make sure that won't be a breaking change for existing apps.

**Comment by sclevine** [comment]
@degrim Sorry if it wasn't obvious by the tag change (unscheduled -> scheduled) -- I prioritized having the team look into this. I'll take it out of our backlog if you'd like to address it instead.

**Comment by degrim** [comment]
@sclevine That's fine with me if you want to have someone on your team take this on.

**Comment by sesmith177** [comment]
@degrim According to the [docs](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/hosting#ordering-importance), the command line parameters will override the environment variable.

It looks like we could set the `ASPNETCORE_URLS` environment variable and remove the `--server.urls` parameter and the apps using `ConfigurationBuilder` would still work

**Comment by degrim** [comment]
@sesmith177 As long as the environment variable works with ASP.NET Core 1.0.0 and above, I think that's probably what we should do then.

**Comment by sesmith177** [comment]
It looks like these have been around since before 1.0; see https://github.com/aspnet/Announcements/issues/108#issuecomment-199967370 and related

**Comment by nathan-alden** [comment]
The good thing about `ASPNETCORE_URLS` is that is _just works_. I don't need to remember to use `ConfigurationBuilder`. It just works in all scenarios.

**Comment by degrim** [comment]
@sesmith177 Sounds good.
@nathan-alden Agreed, using the environment variable instead of command-line parameters will provide a better user experience.

**Comment by sclevine** [comment]
Fix is on develop and will go out with 1.0.15:
https://www.pivotaltracker.com/story/show/142846767
Closing this issue -- follow the Tracker story to monitor the release progress.
